File: clients/ios/PremiumComponents.swift
// PremiumComponents.swift — Phase 3 non-default UI primitives (iOS)

import SwiftUI

// MARK: - Search

struct PremiumSearchField: View {
    @Environment(\.colorScheme) private var colorScheme
    @Binding var text: String
    var placeholder: String = "Search…"

    var body: some View {
        HStack(spacing: SyncTokens.space2) {
            Image(systemName: "magnifyingglass")
                .font(.system(size: 16, weight: .medium))
                .foregroundStyle(SyncTokens.slateMuted)
            TextField(placeholder, text: $text)
                .font(SyncFont.bodySm())
                .foregroundStyle(AppSurfaces.onSurface(colorScheme))
        }
        .padding(.horizontal, SyncTokens.space4)
        .frame(height: 44)
        .background(AppSurfaces.card(colorScheme))
        .clipShape(RoundedRectangle(cornerRadius: SyncTokens.radiusInput))
        .overlay(
            RoundedRectangle(cornerRadius: SyncTokens.radiusInput)
                .stroke(AppSurfaces.cardBorder(colorScheme), lineWidth: 1)
        )
    }
}

// MARK: - Progress

struct PremiumLinearProgress: View {
    var progress: Double

    var body: some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Capsule().fill(SyncTokens.cardBorder)
                Capsule()
                    .fill(
                        LinearGradient(
                            colors: [SyncTokens.teal, SyncTokens.tealLight],
                            startPoint: .leading,
                            endPoint: .trailing
                        )
                    )
                    .frame(width: max(0, geo.size.width * min(1, max(0, progress))))
            }
        }
        .frame(height: 6)
    }
}

// MARK: - Chips

enum PremiumChipVariant { case success, warning, danger, primary, neutral }

struct PremiumChip: View {
    let label: String
    var variant: PremiumChipVariant = .neutral

    private var colors: (bg: Color, fg: Color) {
        switch variant {
        case .success: return (SyncTokens.success.opacity(0.12), SyncTokens.success)
        case .warning: return (SyncTokens.warning.opacity(0.12), SyncTokens.warning)
        case .danger: return (SyncTokens.danger.opacity(0.12), SyncTokens.danger)
        case .primary: return (SyncTokens.teal.opacity(0.12), SyncTokens.teal)
        case .neutral: return (SyncTokens.slateMuted.opacity(0.12), SyncTokens.slateSecondary)
        }
    }

    var body: some View {
        Text(label)
            .font(SyncFont.caption().weight(.semibold))
            .foregroundStyle(colors.fg)
            .padding(.horizontal, 12)
            .padding(.vertical, 6)
            .background(colors.bg)
            .clipShape(Capsule())
    }
}

// MARK: - Skeleton

struct PremiumSkeleton: View {
    var rows: Int = 4
    @State private var phase: CGFloat = 0

    var body: some View {
        VStack(alignment: .leading, spacing: SyncTokens.space3) {
            RoundedRectangle(cornerRadius: SyncTokens.radiusMd)
                .fill(shimmer)
                .frame(width: 140, height: 24)
            ForEach(0..<rows, id: \.self) { _ in
                RoundedRectangle(cornerRadius: SyncTokens.radiusCard)
                    .fill(shimmer)
                    .frame(height: 56)
            }
        }
        .onAppear {
            withAnimation(.linear(duration: 1.4).repeatForever(autoreverses: false)) {
                phase = 1
            }
        }
    }

    private var shimmer: LinearGradient {
        LinearGradient(
            colors: [
                SyncTokens.slateMuted.opacity(0.15),
                SyncTokens.slateMuted.opacity(0.28),
                SyncTokens.slateMuted.opacity(0.15),
            ],
            startPoint: UnitPoint(x: phase - 0.3, y: 0.5),
            endPoint: UnitPoint(x: phase + 0.3, y: 0.5)
        )
    }
}

// MARK: - Bottom sheet

struct PremiumBottomSheet<Content: View>: View {
    @Environment(\.colorScheme) private var colorScheme
    @Binding var visible: Bool
    var title: String?
    var onDismiss: () -> Void
    @ViewBuilder var content: Content

    var body: some View {
        ZStack {
            if visible {
                Color.black.opacity(0.55)
                    .ignoresSafeArea()
                    .onTapGesture { onDismiss() }
                    .transition(.opacity)

                VStack(spacing: 0) {
                    Spacer()
                    VStack(spacing: SyncTokens.space4) {
                        Capsule()
                            .fill(SyncTokens.cardBorder)
                            .frame(width: 36, height: 4)
                        if let title {
                            HStack {
                                Text(title).font(SyncFont.titleLg())
                                Spacer()
                                PremiumIconButton(systemName: "xmark", action: onDismiss)
                            }
                        }
                        content
                    }
                    .padding(SyncTokens.space6)
                    .frame(maxWidth: .infinity)
                    .background(AppSurfaces.card(colorScheme))
                    .clipShape(RoundedRectangle(cornerRadius: SyncTokens.radiusCard, style: .continuous))
                    .overlay(
                        RoundedRectangle(cornerRadius: SyncTokens.radiusCard, style: .continuous)
                            .stroke(AppSurfaces.cardBorder(colorScheme), lineWidth: 1)
                    )
                    .transition(.move(edge: .bottom).combined(with: .opacity))
                }
            }
        }
        .animation(.easeOut(duration: SyncTokens.durationNormal), value: visible)
    }
}

// MARK: - Icon button

struct PremiumIconButton: View {
    @Environment(\.colorScheme) private var colorScheme
    let systemName: String
    var tint: Color = SyncTokens.slateSecondary
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Image(systemName: systemName)
                .font(.system(size: 16, weight: .semibold))
                .foregroundStyle(tint)
                .frame(width: 40, height: 40)
                .background(AppSurfaces.card(colorScheme))
                .clipShape(Circle())
                .overlay(Circle().stroke(AppSurfaces.cardBorder(colorScheme), lineWidth: 1))
        }
        .buttonStyle(.plain)
    }
}

// MARK: - Modal overlay

struct PremiumAppModalOverlay: View {
    @Environment(\.colorScheme) private var colorScheme
    let title: String
    let message: String
    let confirmText: String
    let dismissText: String
    let onConfirm: () -> Void
    let onDismiss: () -> Void

    var body: some View {
        ZStack {
            Color.black.opacity(0.55).ignoresSafeArea().onTapGesture(perform: onDismiss)
            VStack(spacing: SyncTokens.space4) {
                Text(title).font(SyncFont.titleLg()).foregroundStyle(AppSurfaces.onSurface(colorScheme))
                Text(message)
                    .font(SyncFont.bodySm())
                    .foregroundStyle(SyncTokens.slateSecondary)
                    .multilineTextAlignment(.center)
                HStack(spacing: SyncTokens.space3) {
                    GhostButton(title: dismissText, action: onDismiss)
                    PrimaryButton(text: confirmText, action: onConfirm)
                }
            }
            .padding(SyncTokens.space6)
            .frame(maxWidth: 340)
            .background(AppSurfaces.card(colorScheme))
            .clipShape(RoundedRectangle(cornerRadius: SyncTokens.radiusDialog))
            .overlay(
                RoundedRectangle(cornerRadius: SyncTokens.radiusDialog)
                    .stroke(AppSurfaces.cardBorder(colorScheme), lineWidth: 1)
            )
            .shadow(color: .black.opacity(0.2), radius: 24, y: 12)
        }
    }
}

typealias AppSkeleton = PremiumSkeleton

File: clients/android/app/src/main/java/com/syncbridge/android/ui/components/PremiumComponents.kt
package com.syncbridge.android.ui.components

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material.icons.outlined.Search
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import com.syncbridge.android.ui.theme.SyncTokens

// ── Buttons (no Material Button) ─────────────────────────────────────────────

@Composable
fun PremiumPrimaryButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    loading: Boolean = false,
    block: Boolean = true,
) {
    val shape = RoundedCornerShape(SyncTokens.RadiusButton)
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    Box(
        modifier = modifier
            .then(if (block) Modifier.fillMaxWidth() else Modifier)
            .scale(if (pressed) 0.97f else 1f)
            .height(56.dp)
            .clip(shape)
            .background(
                if (enabled && !loading) {
                    Brush.linearGradient(listOf(SyncTokens.Teal, SyncTokens.TealLight))
                } else {
                    Brush.linearGradient(listOf(SyncTokens.Teal.copy(0.4f), SyncTokens.TealLight.copy(0.4f)))
                },
            )
            .clickable(
                enabled = enabled && !loading,
                interactionSource = interaction,
                indication = null,
                onClick = onClick,
            ),
        contentAlignment = Alignment.Center,
    ) {
        if (loading) {
            Row(horizontalArrangement = Arrangement.spacedBy(SyncTokens.Space2), verticalAlignment = Alignment.CenterVertically) {
                CircularProgressIndicator(Modifier.size(20.dp), strokeWidth = 2.dp, color = Color.White)
                Text(text, fontSize = 16.sp, fontWeight = FontWeight.SemiBold, color = Color.White)
            }
        } else {
            Text(text, fontSize = 16.sp, fontWeight = FontWeight.SemiBold, color = Color.White)
        }
    }
}

@Composable
fun PremiumGhostButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(SyncTokens.RadiusSm)
    Box(
        modifier = modifier
            .heightIn(min = 32.dp)
            .clip(shape)
            .border(1.dp, AppSurfaces.cardStroke(), shape)
            .background(Color.Transparent)
            .clickable(onClick = onClick)
            .padding(horizontal = SyncTokens.Space4, vertical = SyncTokens.Space2),
        contentAlignment = Alignment.Center,
    ) {
        Text(text, fontSize = 12.sp, fontWeight = FontWeight.SemiBold, color = SyncTokens.SlateSecondary)
    }
}

@Composable
fun PremiumDangerButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(SyncTokens.RadiusButton)
    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(48.dp)
            .clip(shape)
            .background(SyncTokens.Danger.copy(0.12f))
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Text(text, fontSize = 15.sp, fontWeight = FontWeight.SemiBold, color = SyncTokens.Danger)
    }
}

@Composable
fun PremiumIconButton(
    onClick: () -> Unit,
    icon: ImageVector,
    contentDescription: String?,
    modifier: Modifier = Modifier,
    tint: Color = SyncTokens.SlateSecondary,
) {
    Box(
        modifier = modifier
            .size(40.dp)
            .clip(CircleShape)
            .border(1.dp, AppSurfaces.cardStroke(), CircleShape)
            .background(AppSurfaces.card())
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Icon(icon, contentDescription, tint = tint, modifier = Modifier.size(20.dp))
    }
}

// ── Inputs ────────────────────────────────────────────────────────────────────

@Composable
fun PremiumTextField(
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    placeholder: String = "",
    error: String? = null,
    singleLine: Boolean = true,
    minLines: Int = 1,
    password: Boolean = false,
    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
) {
    val shape = RoundedCornerShape(SyncTokens.RadiusInput)
    val borderColor = if (error != null) SyncTokens.Danger else AppSurfaces.cardStroke()
    Column(modifier = modifier.fillMaxWidth()) {
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            modifier = Modifier
                .fillMaxWidth()
                .clip(shape)
                .border(1.dp, borderColor, shape)
                .background(AppSurfaces.card())
                .padding(horizontal = SyncTokens.Space4, vertical = SyncTokens.Space3),
            textStyle = TextStyle(fontSize = 14.sp, color = SyncTokens.SlateText),
            singleLine = singleLine,
            minLines = minLines,
            visualTransformation = if (password) PasswordVisualTransformation() else VisualTransformation.None,
            keyboardOptions = keyboardOptions,
            cursorBrush = SolidColor(SyncTokens.Teal),
            decorationBox = { inner ->
                Box {
                    if (value.isEmpty() && placeholder.isNotEmpty()) {
                        Text(placeholder, color = SyncTokens.SlateMuted, fontSize = 14.sp)
                    }
                    inner()
                }
            },
        )
        error?.let {
            Text(it, color = SyncTokens.Danger, fontSize = 13.sp, modifier = Modifier.padding(top = SyncTokens.Space2))
        }
    }
}

@Composable
fun PremiumSearchField(
    value: String,
    onValueChange: (String) -> Unit,
    placeholder: String = "Search…",
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(SyncTokens.RadiusContainerSm)
    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(48.dp)
            .clip(shape)
            .border(1.dp, AppSurfaces.cardStroke(), shape)
            .background(AppSurfaces.card())
            .padding(horizontal = SyncTokens.Space4),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(SyncTokens.Space2),
    ) {
        Icon(Icons.Outlined.Search, null, tint = SyncTokens.SlateMuted, modifier = Modifier.size(18.dp))
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            modifier = Modifier.fillMaxWidth(),
            textStyle = TextStyle(fontSize = 14.sp, color = SyncTokens.SlateText),
            singleLine = true,
            cursorBrush = SolidColor(SyncTokens.Teal),
            decorationBox = { inner ->
                Box(Modifier.fillMaxWidth()) {
                    if (value.isEmpty()) Text(placeholder, color = SyncTokens.SlateMuted, fontSize = 14.sp)
                    inner()
                }
            },
        )
    }
}

// ── Dialog (no AlertDialog) ─────────────────────────────────────────────────

@Composable
fun PremiumAppModal(
    title: String,
    message: String,
    confirmText: String,
    dismissText: String,
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
) {
    Dialog(onDismissRequest = onDismiss, properties = DialogProperties(usePlatformDefaultWidth = false)) {
        Box(
            Modifier
                .fillMaxWidth(0.9f)
                .clip(RoundedCornerShape(SyncTokens.RadiusContainerLg))
                .background(AppSurfaces.card())
                .border(1.dp, AppSurfaces.cardStroke(), RoundedCornerShape(SyncTokens.RadiusContainerLg))
                .padding(SyncTokens.Space6),
        ) {
            Column(verticalArrangement = Arrangement.spacedBy(SyncTokens.Space4)) {
                Text(title, fontSize = 20.sp, fontWeight = FontWeight.Bold, color = SyncTokens.SlateText)
                Text(message, fontSize = 14.sp, color = SyncTokens.SlateSecondary, lineHeight = 22.sp)
                Row(horizontalArrangement = Arrangement.spacedBy(SyncTokens.Space3)) {
                    PremiumGhostButton(dismissText, onDismiss, modifier = Modifier.weight(1f))
                    PremiumPrimaryButton(confirmText, onConfirm, modifier = Modifier.weight(1f))
                }
            }
        }
    }
}

// ── Bottom sheet ────────────────────────────────────────────────────────────

@Composable
fun PremiumBottomSheet(
    visible: Boolean,
    title: String?,
    onDismiss: () -> Unit,
    content: @Composable () -> Unit,
) {
    AnimatedVisibility(
        visible = visible,
        enter = fadeIn(tween(SyncTokens.DurationNormal)) + slideInVertically { it },
        exit = fadeOut(tween(SyncTokens.DurationFast)) + slideOutVertically { it },
    ) {
        Box(Modifier.fillMaxSize()) {
            Box(
                Modifier
                    .fillMaxSize()
                    .background(Color.Black.copy(0.55f))
                    .clickable(indication = null, interactionSource = remember { MutableInteractionSource() }) { onDismiss() },
            )
            Column(
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(topStart = SyncTokens.RadiusContainerLg, topEnd = SyncTokens.RadiusContainerLg))
                    .background(AppSurfaces.card())
                    .border(1.dp, AppSurfaces.cardStroke(), RoundedCornerShape(topStart = SyncTokens.RadiusContainerLg, topEnd = SyncTokens.RadiusContainerLg))
                    .padding(SyncTokens.Space6),
            ) {
                Box(
                    Modifier
                        .width(36.dp)
                        .height(4.dp)
                        .clip(CircleShape)
                        .background(SyncTokens.CardBorder)
                        .align(Alignment.CenterHorizontally),
                )
                Spacer(Modifier.height(SyncTokens.Space4))
                title?.let {
                    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween, verticalAlignment = Alignment.CenterVertically) {
                        Text(it, fontSize = 18.sp, fontWeight = FontWeight.SemiBold)
                        PremiumIconButton(onClick = onDismiss, icon = Icons.Outlined.Close, contentDescription = "Close")
                    }
                    Spacer(Modifier.height(SyncTokens.Space3))
                }
                content()
            }
        }
    }
}

// ── Skeleton ────────────────────────────────────────────────────────────────

@Composable
fun PremiumSkeleton(rows: Int = 4, modifier: Modifier = Modifier) {
    val transition = rememberInfiniteTransition(label = "shimmer")
    val offset by transition.animateFloat(
        initialValue = 0f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(tween(1400, easing = LinearEasing), RepeatMode.Restart),
        label = "shimmer",
    )
    Column(modifier, verticalArrangement = Arrangement.spacedBy(SyncTokens.Space3)) {
        Box(
            Modifier
                .fillMaxWidth(0.4f)
                .height(24.dp)
                .clip(RoundedCornerShape(SyncTokens.RadiusMd))
                .background(shimmerBrush(offset)),
        )
        repeat(rows) {
            Box(
                Modifier
                    .fillMaxWidth()
                    .height(56.dp)
                    .clip(RoundedCornerShape(SyncTokens.RadiusCard))
                    .background(shimmerBrush(offset)),
            )
        }
    }
}

private fun shimmerBrush(progress: Float): Brush {
    val base = SyncTokens.SlateMuted.copy(0.15f)
    val highlight = SyncTokens.SlateMuted.copy(0.28f)
    return Brush.horizontalGradient(
        colors = listOf(base, highlight, base),
        startX = progress * 800f,
        endX = progress * 800f + 400f,
    )
}

// ── Progress ────────────────────────────────────────────────────────────────

@Composable
fun PremiumLinearProgress(
    progress: Float,
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(999.dp)
    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(6.dp)
            .clip(shape)
            .background(SyncTokens.CardBorder),
    ) {
        Box(
            Modifier
                .fillMaxHeight()
                .fillMaxWidth(progress.coerceIn(0f, 1f))
                .clip(shape)
                .background(Brush.horizontalGradient(listOf(SyncTokens.Teal, SyncTokens.TealLight))),
        )
    }
}

// ── Chips ───────────────────────────────────────────────────────────────────

enum class ChipVariant { Success, Warning, Danger, Primary, Neutral }

@Composable
fun PremiumChip(
    label: String,
    variant: ChipVariant = ChipVariant.Neutral,
    modifier: Modifier = Modifier,
) {
    val (bg, fg) = when (variant) {
        ChipVariant.Success -> SyncTokens.Success.copy(0.12f) to SyncTokens.Success
        ChipVariant.Warning -> SyncTokens.Warning.copy(0.12f) to SyncTokens.Warning
        ChipVariant.Danger -> SyncTokens.Danger.copy(0.12f) to SyncTokens.Danger
        ChipVariant.Primary -> SyncTokens.Teal.copy(0.12f) to SyncTokens.Teal
        ChipVariant.Neutral -> SyncTokens.SlateMuted.copy(0.12f) to SyncTokens.SlateSecondary
    }
    Box(
        modifier = modifier
            .clip(RoundedCornerShape(SyncTokens.RadiusChip))
            .background(bg)
            .padding(horizontal = 12.dp, vertical = 6.dp),
    ) {
        Text(label, fontSize = 11.sp, fontWeight = FontWeight.SemiBold, color = fg)
    }
}
